### ISSUE TITLE
multiple currency pairs 

### ISSUE THREAD

**Issue by dadrake3** [enigmampc/catalyst#400]
Dear Catalyst Maintainers,

Before I tell you about my issue, let me describe my environment:

# Environment

* Operating System: OsX 10.13.2 (Windows Version or `$ uname --all`)
* Python Version: 3.6.5
* Python Bitness: 64
* How did you install Catalyst: pip 
* Python packages:
aiodns==1.1.1
aiohttp==3.3.2
alembic==0.9.9
async-timeout==3.0.0
attrdict==2.0.0
attrs==18.1.0
bcolz==1.2.1
boto3==1.7.40
botocore==1.10.40
Bottleneck==1.2.1
cchardet==2.1.1
ccxt==1.14.207
certifi==2018.4.16
chardet==3.0.4
click==6.7
contextlib2==0.5.5
cycler==0.10.0
cyordereddict==1.0.0
Cython==0.28.3
cytoolz==0.9.0.1
decorator==4.3.0
docutils==0.14
empyrical==0.2.1
enigma-catalyst==0.5.13
eth-abi==1.1.1
eth-account==0.2.3
eth-hash==0.1.4
eth-keyfile==0.5.1
eth-keys==0.2.0b3
eth-rlp==0.1.2
eth-utils==1.0.3
hexbytes==0.1.0
idna==2.7
idna-ssl==1.0.1
intervaltree==2.1.0
jmespath==0.9.3
kiwisolver==1.0.1
Logbook==1.4.0
lru-dict==1.1.6
lxml==4.2.1
Mako==1.0.7
MarkupSafe==1.0
matplotlib==2.2.2
multidict==4.3.1
multipledispatch==0.5.0
networkx==2.1
numexpr==2.6.5
numpy==1.14.5
pandas==0.19.2
pandas-datareader==0.6.0
parsimonious==0.8.0
patsy==0.5.0
pycares==2.3.0
pycryptodome==3.6.1
pyparsing==2.2.0
python-dateutil==2.7.3
python-editor==1.0.3
pytz==2018.4
redo==1.6
requests==2.19.1
requests-file==1.4.3
requests-ftp==0.3.1
requests-toolbelt==0.8.0
rlp==1.0.1
s3transfer==0.1.13
scipy==1.1.0
six==1.11.0
sortedcontainers==2.0.4
SQLAlchemy==1.2.8
statsmodels==0.9.0
TA-Lib==0.4.17
tables==3.4.4
toolz==0.9.0
urllib3==1.23
web3==4.3.0
websockets==4.0.1
wrapt==1.10.11
yarl==1.1.0

Now that you know a little about me, let me tell you about the issue I am
having:

# Description of Issue
When executing a trade in backtest mode I am trying to first simulate purchasing one btc with usd, then on the next call to handle data I want to use some of that btc to purchase 1 eth on the same exchange. My quote currency is usd which I think may be the problem because cash gets decreased by the price of the eth in btc, which doesn't make sense...  

Inside the context variable what I expect to happen is the first order will decrease cash by ~10,000$, then the second order wont decrease cash but instead will decrease the open position of the 1 btc just purchased by ~0.07. 

The initial values should be 10,000 usd, 0 btc, 0eth, 
then after one iteration 0 usd, 1 btc, 0 eth
then finally 0 usd, 0.93 btc, 1 eth

Here is how you can reproduce this issue on your machine:

## Reproduction Steps

if context.toTrade == 0:
        order(asset=context.btc_usd,
              amount=1)

    elif context.toTrade == 1:
        order(asset=context.eth_btc,
              amount=1)

    context.toTrade += 1

## What steps have you taken to resolve this already?

stepped through the debugger and inspected the context variables for the open position list and cash values

Sincerely,
D

**Comment by AvishaiW** [comment]
Hi, This is a duplicate of #297

**Comment by dadrake3** [comment]
Could you explain why the positions are stored as currency pairs and not as individual currencies. Having a position of 1 usd_btc doesnt make much sense when in reality that means I have 1 btc. Also, when performing live trading do all my quote currencies have to be the same or can I trade arbitrary pairs? The reason I am asking this is because in order to perform arbitrage on more than one currency pair I would have to write an algorithm  for each desired pair which doesn't make much sense eitgher.

**Comment by sam31415** [comment]
@dadrake3 Another related issue is #322. This is all due to the fact that Catalyst was forked from Zipline, which was developed to trade stocks and futures, not currency pairs. I doubt there is a simple fix for that, as it would require rethinking the whole structure of the engine.

**Comment by dadrake3** [comment]
@sam31415 do you know if you can have multiple quote currencies when live trading, like could I trade btc_eth and xrp_usd in one algorithm? This seems like its only an issue of tracking the portfolio balances for back testing / order simulation..

**Comment by sam31415** [comment]
@dadrake3 As far as I know it's not possible. I once tried to have a strategy trade on Binance with USDT as a base currency, but trying to take advantage of the many xxx_btc pairs by trading simultaneously xxx_btc and btc_usdt. It turned out to be impossible.

**Comment by sam31415** [comment]
@dadrake3 I don't think this would even be a solution, because the different algorithms would not be able to share the balances. If Algo 1 sells the btc Algo 2 bought, Algo 2 will complain that its balance went down, adjust it and count it as a pure loss.

I didn't think much about arbitrage strategies so far.

**Comment by dadrake3** [comment]
@sam31415 Can I just pull the current balance off each exchange such that catalyst wont complain?

**Comment by sam31415** [comment]
@dadrake3 Yes you can. But Catalyst records its own balance for each pair, and this balance is impossible to modify as far as I know (#312). My guess is that you'll have trouble ordering if it's off.

**Comment by dadrake3** [comment]
@sam31415 So just to clarify, If i am currently holding USD, BTC, and ETH on binance (in real life), I could not use catalyst to buys BTC with USD, then buy ETH with BTC, and lastly to buy USD with ETH? Or will the orders properly execute but the local portfolio will be out of wack

**Comment by sam31415** [comment]
Certainly not with a single algo. And yes, the local portfolio would be out of wack.
